clientv3/integration/watch_test.go

```diff
@@ -1060,3 +1060,46 @@ func TestWatchCancelDisconnected(t *testing.T) {
 		t.Fatal("took too long to cancel disconnected watcher")
 	}
 }
+
+// TestWatchKeepAlive ensures that watch discovers it cannot talk to server
+// and then switch to another endpoint with keep-alive parameters.
+func TestWatchKeepAlive(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer clus.Terminate(t)
+
+	ccfg := clientv3.Config{
+		Endpoints:            []string{clus.Members[0].GRPCAddr(), clus.Members[1].GRPCAddr()},
+		DialKeepAliveTime:    time.Second,
+		DialKeepAliveTimeout: 500 * time.Millisecond,
+	}
+	cli, err := clientv3.New(ccfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cli.Close()
+
+	wch := cli.Watch(context.Background(), "foo", clientv3.WithCreatedNotify())
+	if _, ok := <-wch; !ok {
+		t.Fatalf("watch failed")
+	}
+	clus.Members[0].Blackhole()
+
+	// expect client to switch endpoints from keepalive ping
+	time.Sleep(3 * ccfg.DialKeepAliveTime)
+
+	donec := make(chan struct{})
+	go func() {
+		defer close(donec)
+		if _, err = cli.Put(context.TODO(), "foo", "bar"); err != nil {
```

Comment:
`clus.Client(1).Put(...)` and the test will fail on master as expected. The Put acts like a keepalive on the grpc connection.